### PR TITLE
chore(website): improve error handling in playground

### PR DESCRIPTION
**Pull request by armano2**
## PR Checklist

-   [x] Addresses an existing open issue: fixes #000
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

- correct bug with event missing in monaco for  ts <= 4.0.5 https://github.com/microsoft/monaco-editor/issues/313
  with all those new features, editor just crashes for old monaco versions
- ensure that all proper libs are download for specified typescript versions
  easiest way to test is to check if editor recognize `Promise` as valid type, [Test 1]
- reduce computation complexity when updating codeFixers
- In case of fatal parsing error fallback to error list, this unifies how errors are displayed between screens
  Old:
  image

  New: 
  image
- typescript errors markers are going to be now printed only for user provided file with exclusion to externally provided files
  ```ts
  import React from 'react'
  ```
  Old:
  image

  New:
  image

[Test 1]: https://deploy-preview-5083--typescript-eslint.netlify.app/play/#ts=3.3.3&sourceType=module&code=PQKgsAUApgzgNgSwHYBdIAJPoEQAEUCeADrAMYBOCRKAtLIqsEgPY0BmczAhisgOY0i5ZgFsEMWNgBc6ANoByKOWHl5AXQA0GLHkIkYFKrXrIUTVmJgBXCQBNBwy5JkKlK9VohYc+YmUrUdPCm5jRWSDZQ9gBuXOQw0nKKysyqmtqYun4GAcbBjDAolKS0AEbMzHBQXEh0AB5CsDAIzEgJLjhuqdga6ADeOFxwnADuAMpF-IlsQxK92EOjAHJWIqVK07NQ84vMIyvDXKVVAPKlAFZQJZtwEugAvmqQIMCQkAhs6AAURVZQ6AAyAHoACM6AA1KCAJT9e6QEYACwQVW+M1u-wAPhjBtgIThStgYX04RAvnUZMwLlcUDCALwAPnQv0x2LRdyx6Dq6AA-EzyH90DI2VAANxvCDRZgIWzoJBQEboAAKjnEUC+XzpjIG9yhIvQwGA6EAMuSy1gcbi8JACISiVUwdDiqANVIodBscIlFpIdAUaooKDK20SDX9bTkKAoKzkb1yhWBpzqzX9B66-WGwCg5Ohw5Ho1EHRASZBfTwAyrg6mDcbTexODx+A4g7BHc7yK73UhPa10IUrGw2CCQ31tKRWoV0ER0LTZfKlWW1SGGcmdXrKyaWGEInYaLF4pBC9AW66uDACB23R7eF2e32AEyD7RcEZcBCu2OzxuJydalMrjPoR-Pv6th7s2RAuueHaXt6kTxqqXxEDIsESAAPOEADWLAjEg9JEvcaZVuu4SRDEcQwOKMFzl8b5IfOSbalCFZ-kQx52OKkBOmBrb-ieZ7tp23qLPeXhYCObSugAJFwk4+uGJY0Rqv7oJmLEIHwcrAcJmCiWO4mlNJxb+vJjFKdxzRqVE4reNprqyFwvSlGo0kAS+75OAAdIJsiSb0ulqMZmbMTAdhMswrmqh5wwahk+FrqwRFbjuZEFmxEDvJ8XwGaWH4MbCMXVpYxENk4SWIsi-wZbJhmUTlxJ5euBVbjaxWQBVfpZQmMK8sygpulsxmxTQDVREVdotZlRnoByzL9fl4iFU1o2kgAhON1UzfVc2NXOSWQLYVxwHE-x8VB-6kKQUDUDAABqUq2AAwgAQhlpQyAujKStKuqQFwZ0XSg123Y9rVyXO60WJtw0LRIMBAA&rules=N4XyA&tsConfig=N4XyA

### CODE REVIEW

File: packages/website/src/components/editor/LoadedEditor.tsx [hunk 1]
```diff
@@ -56,33 +58,29 @@ export const LoadedEditor: React.FC<LoadedEditorProps> = ({
...
       );
 
-      if (fatalMessage) {
-        setDecorations(
-          sandboxInstance.editor.deltaDecorations(decorations, []),
+      // fallback when event is not preset, ts < 4.0.5
+      if (!sandboxInstance.monaco.editor.onDidChangeMarkers) {
+        const markers = sandboxInstance.monaco.editor.getModelMarkers({});
+        onMarkersChange(
+          parseMarkers(markers, codeActions, sandboxInstance.editor),
```
Comment (armano2) [inline packages/website/src/components/editor/LoadedEditor.tsx:77]: this is not ideal, as we in some cases are not going to retrieve typescript errors, but to my knowledge there is no alternative if we can't use `onDidChangeMarkers`

we may not see ts errors if monaco ts webworker is still processing data, as we are updating this only after linting is done

File: packages/website/src/components/editor/useSandboxServices.ts [hunk 2]
```diff
@@ -76,8 +76,14 @@ export const useSandboxServices = (
           colorMode === 'dark' ? 'vs-dark' : 'vs-light',
         );
 
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
+        const libs = ((window.ts as any).libs as string[]) ?? ['esnext'];
```
Comment (armano2) [inline packages/website/src/components/editor/useSandboxServices.ts:80]: `libs` is not exposed publicly, but we need a way to retrieve list of libs that are valid for specified typescript version

Comment (armano2) [reply packages/website/src/components/editor/useSandboxServices.ts:80]: this change makes this code backward and forward compatible as long as ts is going to have this field exposed

Comment (armano2) [reply packages/website/src/components/editor/useSandboxServices.ts:80]: alternatively we could define list of libs that should be allowed per ts version, but its easy to make mistake in there

**Comment by bradzacher** [comment]
Question:
do we really care about versions of TS that old for our playground? TSv4 was released Aug 2020 (almost 2 years ago).
If we need to maintain these sorts of hacks to make it work, it might not be worth the effort?
I haven't searched - but I suspect most people just use the latest version of TS when using the playground.

**Comment by bradzacher** [comment]
> this list actually already limited to version supported by us

Yeah, sorry - I know we support older than v4.1 (as much as I hate that we do..!)
I mean specifically for the playground - do we care about allowing those older versions?

I kind-of feel like we could just say "stuff it minimum v4.1" on the playground.
The intersection between people reporting bugs with our tooling and people using a 2y old version of TS is *probably* small-to-non-existent?

**Comment by armano2** [comment]
I'm using it sometimes to check if there is no regressions in rules for older version of typescript, and its actually easier to do so in browser than creating new project

**Comment by armano2** [comment]
i found one more issue, in case if user decides to import external library, that has type definitions and this library has some warnings/errors for specified typescript version, that makes editor print lots of errors in the list and make list of errors unusuable,

from now on we are going to limit displaying errors to only one that are from current model (user provided code)

**Pull request closed without merging**